beacon-chain/simulator/service.go

```diff
@@ -0,0 +1,122 @@
+package simulator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"golang.org/x/crypto/blake2b"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/types"
+	"github.com/prysmaticlabs/prysm/shared/p2p"
+
+	pb "github.com/prysmaticlabs/prysm/proto/sharding/v1"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithField("prefix", "simulator")
+
+// Simulator struct.
+type Simulator struct {
+	ctx                    context.Context
+	cancel                 context.CancelFunc
+	p2p                    types.P2P
+	web3Service            types.POWChainService
+	delay                  time.Duration
+	broadcastedBlockHashes map[[32]byte]*types.Block
+	blockRequestChan       chan p2p.Message
+}
+
+// Config options for the simulator service.
+type Config struct {
+	Delay           time.Duration
+	BlockRequestBuf int
+}
+
+// DefaultConfig options for the simulator.
+func DefaultConfig() *Config {
+	return &Config{Delay: time.Second * 7, BlockRequestBuf: 100}
+}
+
+// NewSimulator hi.
+func NewSimulator(ctx context.Context, cfg *Config, beaconp2p types.P2P, web3Service types.POWChainService) *Simulator {
+	ctx, cancel := context.WithCancel(ctx)
+	return &Simulator{
+		ctx:         ctx,
+		cancel:      cancel,
+		p2p:         beaconp2p,
+		web3Service: web3Service,
+		delay:       cfg.Delay,
+		broadcastedBlockHashes: make(map[[32]byte]*types.Block),
+		blockRequestChan:       make(chan p2p.Message, cfg.BlockRequestBuf),
+	}
+}
+
+// Start the sim.
+func (sim *Simulator) Start() {
+	log.Info("Starting service")
+	go sim.run(time.NewTicker(sim.delay).C, sim.ctx.Done())
+}
+
+// Stop the sim.
+func (sim *Simulator) Stop() error {
+	defer sim.cancel()
+	log.Info("Stopping service")
+	return nil
+}
+
+func (sim *Simulator) run(delayChan <-chan time.Time, done <-chan struct{}) {
+	blockReqSub := sim.p2p.Feed(pb.BeaconBlockRequest{}).Subscribe(sim.blockRequestChan)
```

Comment:
should be able to use `sim.p2p.Subscribe(pb.BeaconBlockRequest{}, sim.blockRequestChan)` now

Reply:
+1 use the that new subscribe method.